## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [696]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [697]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [698]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [699]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [700]:
# What three item categories had the most entries?
value_counts = df['category'].value_counts()
top_3_categories = value_counts.nlargest(3)
top_3_categories.head()


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [701]:
# For the category with the most entries,
# which subcategory had the most entries?
top_subcategory = df['subcategory'].value_counts().sort_values(ascending=False)
top_subcategory.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [702]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().sort_values(ascending=False)
top_clients.head()


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [703]:
# Store the client ids of those top 5 clients in a list.
top_5_clients_id_list = top_5_clients.nlargest(5).index.tolist()
print("Top 5 Client IDs are:",top_5_clients_id_list)


Top 5 Client IDs are: [78965, 86721, 30143, 85931, 77354]


In [704]:
# How many total units (the qty column) did the
# client with the most entries order order?
most_orders_client_id = df['client_id'].value_counts().idxmax()
client_orders = df[df['client_id'] == most_orders_client_id]
total_units = client_orders['qty'].sum()
print("Client ID with most entries:", most_orders_client_id)
print("Total units ordered:", total_units)

Client ID with most entries: 33615
Total units ordered: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [705]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df = df.assign(subtotal=lambda df:
              (df.unit_price * df.qty))
df.head(10) 


first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  \
0            wall art     1096.80     762.71         7.50  105            1   
1                pens       24.95      15.09         1.49   21            0   
2  project management       13.52       7.86         1.68   39            6   
3                pens       36.42      24.85         1.23   29            3   
4                misc      195.10     108.17        46.43   20            1   
5              plants      172.84     154.44         2.79   78            0   
6    kitchen supplies      245.56     231.95         1.73   56            5   
7                pens       20.99      16.97         5.12   22            4   
8      paper products       83.86      59.00         0.27   49            3   
9             seating       66.88      54.15         2.47   78            1   

    subtotal  
0  115164.00  
1     523.95  
2     527.28  
3    1056.18  
4    3902.00  
5   13481.52  
6   13751.36  
7     461.78  
8    4109.14  
9    5216.64

In [706]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def calculate_shipping_price(unit_weight):
  if unit_weight > 50:
    return unit_weight* 7
  else:
    return unit_weight* 10
df['shipping_price'] = df['qty'].apply(calculate_shipping_price)
df['line_price'] = df['subtotal'] + df['shipping_price'] 
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00             735   
1      15.09         1.49   21            0     523.95             210   
2       7.86         1.68   39            6     527.28             390   
3      24.85         1.23   29            3    1056.18             290   
4     108.17        46.43   20            1    3902.00             200   

   line_price  
0   115899.00  
1      733.95  
2      917.28  
3     1346.18  
4     4102.00  

[5 rows x 21 columns]

In [707]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
#df['sales_tax'] = df['subtotal'] * 0.0925
df['total_price'] = (df['subtotal'] + df['shipping_price']) * (1 + 0.0925)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price  line_price  \
0         7.50  105            1  115164.00             735   115899.00   
1         1.49   21            0     523.95             210      733.95   
2         1.68   39            6     527.28             390      917.28   
3         1.23   29            3    1056.18             290     1346.18   
4        46.43   20            1    3902.00             200     4102.00   

     total_price  
0  126619.657500  
1     801.840375  
2    1002.128400  
3    1470.701650  
4    4481.435000  

[5 rows x 22 columns]

In [708]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty']
df['total_cost'] = df['line_cost'] + df['shipping_price']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  shipping_price  line_price    total_price  line_cost  total_cost  
0  115164.00             735   115899.00  126619.657500   80084.55    80819.55  
1     523.95             210      733.95     801.840375     316.89      526.89  
2     527.28             390      917.28    1002.128400     306.54      696.54  
3    1056.18             290     1346.18    1470.701650     720.65     1010.65  
4    3902.00             200     4102.00    4481.435000    2163.40     2363.40  

[5 rows x 24 columns]

In [709]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['total_cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number   subtotal  \
0          17        2023  ...        7.50  105           1  115164.00   
1          20        2023  ...        1.49   21           0     523.95   
2           4        2023  ...        1.68   39           6     527.28   
3          17        2023  ...        1.23   29           3    1056.18   
4          21        2023  ...       46.43   20           1    3902.00   

   shipping_price  line_price    total_price  line_cost  total_cost  \
0             735   115899.00  126619.657500   80084.55    80819.55   
1             210      733.95     801.840375     316.89      526.89   
2             390      917.28    1002.128400     306.54      696.54   
3             290     1346.18    1470.701650     720.65     1010.65   
4             200     4102.00    4481.435000    2163.40     2363.40   

         profit  
0  45800.107500  
1    274.950375  
2    305.588400  
3    460.051650  
4   2118.035000  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [718]:
# Check your work using the totals above
total_orders_df = df.groupby('order_id').agg({
   'total_price': 'sum',
})
order_ids = [2742071,2173913,6128929]
total_orders_df[["total_price"]]

def verify_order_price(order_id):
  order_lines_df = df.loc[df['order_id'] == order_id, 'total_price']
  return round(order_lines_df.sum(),2)

for order_id in order_ids:
  print(f"Order ID {order_id} had a total price of ${verify_order_price(order_id)}")

Order ID 2742071 had a total price of $143312.33
Order ID 2173913 had a total price of $129190.05
Order ID 6128929 had a total price of $787923.31


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [711]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_clients = df.groupby('client_id')['qty'].sum()
top_clients = df.sort_values(by='qty', ascending=False)
top_clients.head(5)


first      last                                  job  \
21801    Julie  Anderson                  Exhibition designer   
25972      Amy   Stewart            Psychologist, educational   
8761   Randall    Watson                    Ceramics designer   
15112   Dennis   Hawkins  Social research officer, government   
50048   Amanda   Burnett                      Records manager   

                       phone                                email  client_id  \
21801             5933099462        anderson_better_4704@bell.biz      78965   
25972     979.793.4884x57205               stewartamy@francis.com      86721   
8761              3478635175  watson.hospital@scott-mcdonald.info      30143   
15112  001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
50048     (959)858-6180x6708            evidence.amanda@clark.biz      77354   

       order_id  order_date  order_week  order_year  ... unit_weight      qty  \
21801   3833568  2023-04-29          17        2023  ...        0.09  3958244   
25972   6057452  2023-03-21          12        2023  ...        9.22   977121   
8761    3566639  2023-02-21           8        2023  ...        1.49   842581   
15112   7263857  2023-03-05           9        2023  ...        2.21   637318   
50048   3089707  2023-04-17          16        2023  ...        6.64   494692   

      line_number      subtotal  shipping_price    line_price   total_price  \
21801           1  5.844743e+08        27707708  6.121820e+08  6.688089e+08   
25972           3  9.810295e+07         6839847  1.049428e+08  1.146500e+08   
8761            2  2.102240e+07         5898067  2.692046e+07  2.941061e+07   
15112           3  2.061214e+08         4461226  2.105826e+08  2.300615e+08   
50048           1  4.018383e+07         3462844  4.364668e+07  4.768399e+07   

          line_cost    total_cost        profit  
21801  3.962598e+08  4.239675e+08  2.448413e+08  
25972  7.137869e+07  7.821854e+07  3.643147e+07  
8761   1.271455e+07  1.861261e+07  1.079799e+07  
15112  1.171645e+08  1.216258e+08  1.084357e+08  
50048  3.424258e+07  3.770542e+07  9.978568e+06  

[5 rows x 25 columns]

In [712]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_df = df[df['client_id'].isin(top_clients.index)]
summary_df = summary_df.groupby('client_id').agg({
  'qty': 'sum',
  'shipping_price': 'sum',
  'total_price': 'sum',
  'profit': 'sum'
})
total_revenue = summary_df['total_price'].sum()

# Sort by total profit (descending order)
summary_df = summary_df.sort_values(by='profit', ascending=False)
summary_df.head(5)


qty  shipping_price   total_price        profit
client_id                                                    
38473      402770         2824367  1.818410e+08  4.320681e+07
24741      239862         1685862  7.851005e+07  3.626174e+07
40457      210792         1477617  5.312902e+07  2.013850e+07
51561       38298          268965  3.722698e+07  1.377582e+07
44673      118251          832008  2.788866e+07  1.333607e+07

In [713]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
clients_df = summary_df.rename(columns={"client_id":"Client ID",
                                        "qty":"Quantity",
                                        "shipping_price":"Shipping Price",
                                        "total_price":"Total Price",
                                        "profit":"Profit", })
clients_df.head()
def currency_in_millions(shipping_price, total_price, profit):
        return f"{'shipping_price'/1000000:.2f}M"
        return f"{'total_price'/1000000:.2f}M"



In [714]:
# Sort the updated data by "Total Profit" form highest to lowest
clients_df = clients_df.sort_values(by='Profit', ascending=False)
#df.assign(Profit=lambda df:(df.Total Price * df.Quantity))
clients_df.head()

Quantity  Shipping Price   Total Price        Profit
client_id                                                      
38473        402770         2824367  1.818410e+08  4.320681e+07
24741        239862         1685862  7.851005e+07  3.626174e+07
40457        210792         1477617  5.312902e+07  2.013850e+07
51561         38298          268965  3.722698e+07  1.377582e+07
44673        118251          832008  2.788866e+07  1.333607e+07